### PR TITLE
Improve support of integer keys in data type Map

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve support of integer keys in data type `Map`.

### CODE REVIEW

File: src/Parsers/ExpressionElementParsers.cpp [hunk 1]
```diff
@@ -1998,7 +1998,6 @@ bool ParserExpressionElement::parseImpl(Pos & pos, ASTPtr & node, Expected & exp
 {
     return ParserSubquery().parse(pos, node, expected)
         || ParserTupleOfLiterals().parse(pos, node, expected)
-        || ParserMapOfLiterals().parse(pos, node, expected)
```
Comment (nikitamikhaylov) [inline src/Parsers/ExpressionElementParsers.cpp:2001]: Why? Does this code enable when I query ` select map(1, 2)` ?

**Review by nikitamikhaylov — APPROVED** [review]
Also there are some questions

File: src/DataTypes/DataTypeMap.cpp [hunk 2]
```diff
@@ -38,14 +39,28 @@ DataTypeMap::DataTypeMap(const DataTypes & elems_)
...
     , nested(std::make_shared<DataTypeArray>(
-        std::make_shared<DataTypeTuple>(DataTypes{key_type_, value_type_}, Names{"keys", "values"}))) {}
+        std::make_shared<DataTypeTuple>(DataTypes{key_type_, value_type_}, Names{"keys", "values"})))
+{
+    assertKeyType();
+}
+
+void DataTypeMap::assertKeyType() const
+{
+    if (!key_type->isValueRepresentedByInteger() && !isStringOrFixedString(*key_type) && !WhichDataType(key_type).isNothing())
```
[the remark this answers is not in the document]
Comment (CurtizJ) [reply src/DataTypes/DataTypeMap.cpp:58]: To allow creating of empty maps for queries like this `select count() from t where m != map()`

File: src/DataTypes/getLeastSupertype.cpp [hunk 3]
```diff
@@ -162,6 +163,36 @@ DataTypePtr getLeastSupertype(const DataTypes & types)
         }
     }
 
+    /// For maps
```
[the remark this answers is not in the document]
Comment (CurtizJ) [reply src/DataTypes/getLeastSupertype.cpp:166]: Yes, it's needed for comparisons.

**Pull request merged**